consensus/src/processes/dagtraversalmanager.rs

```diff
@@ -0,0 +1,147 @@
+use std::{cmp::Reverse, collections::BinaryHeap, sync::Arc};
+
+use crate::{
+    model::stores::{
+        block_window_cache::BlockWindowCacheReader,
+        ghostdag::{GhostdagData, GhostdagStoreReader},
+    },
+    processes::ghostdag::ordering::SortableBlock,
+};
+use consensus_core::blockhash::BlockHashExtensions;
+use hashes::Hash;
+use misc::uint256::Uint256;
+
+#[derive(Clone)]
+pub struct DagTraversalManager<T: GhostdagStoreReader, U: BlockWindowCacheReader> {
+    genesis_hash: Hash,
+    ghostdag_store: Arc<T>,
+    block_window_cache_store: Arc<U>,
+    block_window_cache_store_window_size: usize,
+}
+
+impl<T: GhostdagStoreReader, U: BlockWindowCacheReader> DagTraversalManager<T, U> {
+    pub fn new(
+        genesis_hash: Hash, ghostdag_store: Arc<T>, block_window_cache_store: Arc<U>,
+        block_window_cache_store_window_size: usize,
+    ) -> Self {
+        Self { genesis_hash, ghostdag_store, block_window_cache_store, block_window_cache_store_window_size }
+    }
+    pub fn block_window(
+        &self, high_ghostdag_data: Arc<GhostdagData>, window_size: usize,
+    ) -> BinaryHeap<Reverse<SortableBlock>> {
+        let mut window_heap = SizedUpBlockHeap::new(self.ghostdag_store.clone(), window_size);
+        if window_size == 0 {
+            return window_heap.binary_heap;
+        }
+
+        let mut current_gd = high_ghostdag_data;
+
+        if window_size == self.block_window_cache_store_window_size {
+            if let Some(select_parent_binary_heap) = self
+                .block_window_cache_store
+                .get(&current_gd.selected_parent)
+            {
+                let mut window_heap = SizedUpBlockHeap::from_binary_heap(
+                    self.ghostdag_store.clone(),
+                    window_size,
+                    (*select_parent_binary_heap).clone(),
+                );
+                if current_gd.selected_parent != self.genesis_hash {
+                    self.try_push_mergeset(&mut window_heap, &current_gd);
+                }
+
+                return window_heap.binary_heap;
+            }
+        }
+
+        // Walk down the chain until we finish
+        loop {
+            if current_gd.selected_parent == self.genesis_hash || current_gd.selected_parent.is_origin() {
```

Comment:
if we reach origin this is probably an error, bcs we are expected to validate blocks which have a full window above origin